addons/misra.py

```diff
@@ -256,14 +256,50 @@ def getForLoopExpressions(forToken):
             lpar.astOperand2.astOperand2.astOperand2]
 
 
-def hasFloatComparison(expr):
-    if not expr:
+def findCounterTokens(cond):
+    if not cond:
+        return []
+    if cond.str in ['&&', '||']:
+        c = findCounterTokens(cond.astOperand1)
+        c.extend(findCounterTokens(cond.astOperand2))
+        return c
+    ret = []
+    if ((cond.isArithmeticalOp and cond.astOperand1 and cond.astOperand2) or
+            (cond.isComparisonOp and cond.astOperand1 and cond.astOperand2)):
+        if cond.astOperand1.isName:
+            ret.append(cond.astOperand1)
+        if cond.astOperand2.isName:
+            ret.append(cond.astOperand2)
+        if cond.astOperand1.isOp:
+            ret.extend(findCounterTokens(cond.astOperand1))
+        if cond.astOperand2.isOp:
+            ret.extend(findCounterTokens(cond.astOperand2))
+    return ret
+
+
+def isFloatCounterInWhileLoop(whileToken):
+    if not simpleMatch(whileToken, 'while ('):
+        return False
+    lpar = whileToken.next
+    rpar = lpar.link
+    counterTokens = findCounterTokens(lpar.astOperand2)
+    whileBodyStart = None
+    if simpleMatch(rpar, ') {'):
+        whileBodyStart = rpar.next
+    elif simpleMatch(whileToken.previous, '} while') and simpleMatch(whileToken.previous.link.previous, 'do {'):
+        whileBodyStart = whileToken.previous.link
+    else:
         return False
-    if expr.isLogicalOp:
-        return hasFloatComparison(expr.astOperand1) or hasFloatComparison(expr.astOperand2)
-    if expr.isComparisonOp:
-        # TODO: Use ValueType
-        return cppcheckdata.astIsFloat(expr.astOperand1) or cppcheckdata.astIsFloat(expr.astOperand2)
+    token = whileBodyStart
+    while (token != whileBodyStart.link):
+        token = token.next
+        for counter_str in counterTokens:
```

Comment:
the name `counter_str` is confusing to me. Since it's not some string. how about `counter` or `counterToken`.